### ISSUE TITLE
XF86ScreenSaver not working

### ISSUE THREAD

**Issue by Ratakor** [riverwm/river#935]
When I map XF86ScreenSaver with `riverctl map normal None XF86ScreenSaver spawn 'swaylock'` it is binded without error and I assume that it is catched by river because it disappear from `xev` output but when I click on it, nothing happen. Here is the log:
```
The XKEYBOARD keymap compiler (xkbcomp) reports:
> Warning:          Unsupported maximum keycode 708, clipping.
>                   X11 cannot support keycodes above 255.
> Warning:          Could not resolve keysym XF86CameraAccessEnable
> Warning:          Could not resolve keysym XF86CameraAccessDisable
> Warning:          Could not resolve keysym XF86CameraAccessToggle
> Warning:          Could not resolve keysym XF86NextElement
> Warning:          Could not resolve keysym XF86PreviousElement
> Warning:          Could not resolve keysym XF86AutopilotEngageToggle
> Warning:          Could not resolve keysym XF86MarkWaypoint
> Warning:          Could not resolve keysym XF86Sos
> Warning:          Could not resolve keysym XF86NavChart
> Warning:          Could not resolve keysym XF86FishingChart
> Warning:          Could not resolve keysym XF86SingleRangeRadar
> Warning:          Could not resolve keysym XF86DualRangeRadar
> Warning:          Could not resolve keysym XF86RadarOverlay
> Warning:          Could not resolve keysym XF86TraditionalSonar
> Warning:          Could not resolve keysym XF86ClearvuSonar
> Warning:          Could not resolve keysym XF86SidevuSonar
> Warning:          Could not resolve keysym XF86NavInfo
Errors from xkbcomp are not fatal to the X server
debug(wlroots): [xwayland/xwm.c:1549] unhandled X11 event: MappingNotify (34)
debug(wlroots): [xwayland/xwm.c:1549] unhandled X11 event: MappingNotify (34)
warn: key-binding.c:337: Shift+Control+plus: combo with both explicit modifier and shifted symbol (level=1, mod-mask=0x00000001), replacing with equal
The XKEYBOARD keymap compiler (xkbcomp) reports:
> Warning:          Unsupported maximum keycode 708, clipping.
>                   X11 cannot support keycodes above 255.
> Warning:          Could not resolve keysym XF86CameraAccessEnable
> Warning:          Could not resolve keysym XF86CameraAccessDisable
> Warning:          Could not resolve keysym XF86CameraAccessToggle
> Warning:          Could not resolve keysym XF86NextElement
> Warning:          Could not resolve keysym XF86PreviousElement
> Warning:          Could not resolve keysym XF86AutopilotEngageToggle
> Warning:          Could not resolve keysym XF86MarkWaypoint
> Warning:          Could not resolve keysym XF86Sos
> Warning:          Could not resolve keysym XF86NavChart
> Warning:          Could not resolve keysym XF86FishingChart
> Warning:          Could not resolve keysym XF86SingleRangeRadar
> Warning:          Could not resolve keysym XF86DualRangeRadar
> Warning:          Could not resolve keysym XF86RadarOverlay
> Warning:          Could not resolve keysym XF86TraditionalSonar
> Warning:          Could not resolve keysym XF86ClearvuSonar
> Warning:          Could not resolve keysym XF86SidevuSonar
> Warning:          Could not resolve keysym XF86NavInfo
Errors from xkbcomp are not fatal to the X server
debug(wlroots): [xwayland/xwm.c:1549] unhandled X11 event: MappingNotify (34)
debug(wlroots): [xwayland/xwm.c:1549] unhandled X11 event: MappingNotify (34)
warn: key-binding.c:337: Shift+Control+plus: combo with both explicit modifier and shifted symbol (level=1, mod-mask=0x00000001), replacing with equal
```

**Comment by Ratakor** [comment]
If something similar happens to someone else, don't forget to add `spawn` before the command ;)

**Comment by Leon-Plickat** [comment]
Make sure swaylock actually works by launching it from a terminal.
You can check whether the bind works by mapping a different command to the key for testing.

**Comment by Ratakor** [comment]
Swaylock works it was just an example, the command doesn't matter.

**Comment by ifreund** [comment]
What happens if you use the same command you are trying to map to `XF86ScreenSaver` with another key?

Honestly, I suspect that swaylock isn't the $PATH river was launched with or something similar. `XF86ScreenSaver` disappearing from xev/wev output indicates that the keybinding was successfully created and is being triggered. I think it is far more likely that whatever command you have mapped to it is failing than this particular key being special somehow

**Comment by ifreund** [comment]
Alright, so I've finally figured out what's happening here. libxkbcommon defines the following two kesyms
```c
#define XKB_KEY_XF86Screensaver              0x10081245      /* v3.16   KEY_SCREENSAVER */
```
```c
#define XKB_KEY_XF86ScreenSaver           0x1008ff2d  /* Invoke screensaver         */
```
Your keyboard sends a keycode that gets translated to the `XKB_KEY_XF86ScreenSaver` keysym with value `0x1008ff2d` by your keymap.

However, when the `riverctl map normal None XKB_KEY_XF86ScreenSaver ...` command is used to create a mapping it calls the `xkb_keysym_from_name()` function with the `XKB_KEYSYM_CASE_INSENSITIVE` flag and the value `0x10081245` is returned since it is the lower of the two.

This means that the mapping is never actually triggered.

There's one thing I'm confused about here, you stated that the `XKB_KEY_XF86ScreenSaver` keysym disappears from xev output when mapped but I did not find that to be the case in my testing unless I patched river to use a case sensitive match. Are you certain about that detail?

**Comment by ifreund** [comment]
As for how this issue should be fixed, I'm not sure about that either. The obvious thing to do would be to stop using the case-insensitive keysym matching when creating mappings but that would be a highly breaking change and likely affect all users. For that reason, I don't really want to do that at this time.

An alternative would be to add a `-case-sensitive-keysym` flag or similar to the `riverctl map` command but that leaves this footgun you've stumbled upon unsolved.

**Comment by ifreund** [comment]
Should be fixed by 540ca043df650568bdee4d00b933ed6a13b5de70
